File: day06/python/part1.py
#!/usr/bin/env python3

"""
It must work even if COM is not in the first line.
`input.txt` is like that, so I took the example
and randomized its lines to make the example
similar to the real input.
"""

import helper

from typing import Dict, List


def main() -> None:
    # lines = helper.read_lines("example.txt")
    lines = helper.read_lines("input.txt")

    full: Dict[str, List[str]] = {}
    orbits: Dict[str, int] = {
        'COM': 0    # init.
    }
    for line in lines:
        left, right = line.split(')')
        if left not in full:
            full[left] = []
        full[left].append(right)

    while len(full) > 0:
        for k in full:
            if k in orbits:
                orbit_value = orbits[k]
                reachable = full[k]
                for obj in reachable:
                    orbits[obj] = orbit_value + 1
                del full[k]
                break
            #
        #
    #
    result = sum(orbits.values())
    print(result)

##############################################################################

if __name__ == "__main__":
    main()

File: day06/rust/src/bin/part1.rs
use day06::helper::read_input_file;

use fxhash::FxHashMap as HashMap;

// const INPUT_FILE: &str = "example.txt";
const INPUT_FILE: &str = "input.txt";


fn main()
{
    let lines: Vec<String> = read_input_file(INPUT_FILE);
    //for line in lines.iter() {
    //    println!("{}", line);
    //}

    let mut full: HashMap<&str, Vec<&str>> = HashMap::default();
    let mut orbits: HashMap<&str, i32> = HashMap::default();
    orbits.insert("COM", 0);

    for line in lines.iter() {
        let parts: Vec<&str> = line.split(")").collect();
        let (left, right) = (parts[0], parts[1]);
        if !full.contains_key(left) {
            full.insert(left, Vec::new());
        }
        full.get_mut(left).unwrap().push(right);
    }

    //println!("{:?}", full);

    while full.len() > 0
    {
        for k in full.keys().map(|s| s.clone())
        {
            if orbits.contains_key(k)
            {
                let orbit_value = orbits[k];
                let reachable = &full[k];
                for obj in reachable {
                    orbits.insert(obj, orbit_value + 1);
                }
                full.remove(k);
                break;
            }
        }
    }
    // println!("{:?}", orbits);

    let result: i32 = orbits.values().sum();
    println!("{}", result);
}
